### ISSUE TITLE
System.ArgumentException when trying to invoke .NET method from JavaScript.

### ISSUE THREAD

**Issue by BrunoBlanes** [dotnet/aspnetcore#20293]
### The Bug
Whenever I try to invoke a .NET method from JavaScript, in Blazor WebAssembly, an exception si thrown indicating that the method or assembly was not found.

### To Reproduce
If needed I can provide the full solution, but what I can say is that with Blazor Server I didn't find this issue, only with Blazor WebAssembly.

### Index.razor
``` razor
@page "/"

@using Covid19Tracker.Client.Services
@using Covid19Tracker.Shared.Models
@using Microsoft.JSInterop
@using System.Drawing

@implements IDisposable
@inject IJSRuntime JsRuntime
@inject HttpClient HttpClient
@inject BrowserService BrowserService

<h1>Map</h1>
<div id="tooltip" display="none" style="position: absolute; display: none;">@tooltipContent</div>
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 191000 278400" height="@svgMapHeight">
<g id="StateBorder">
<path id="Espírito Santo" @onmouseover="MouseOver" fill="green" path="[...]">
</g>
</svg>

<style>
    #tooltip {
        background: cornsilk;
        border: 1px solid black;
        border-radius: 5px;
        padding: 5px;
    }
</style>

@code {
    private int svgMapHeight = 0;
    private string tooltipContent;
    public int InnerWidth { get; set; }
    public int InnerHeight { get; set; }

    protected override void OnInitialized()
    {
        BrowserService.OnMouseOver += MouseOverCounty;
        BrowserService.OnResize += BrowserHasResizedAsync;
    }

    private void MouseOverCounty(object sender, EventArgs e)
    {
        tooltipContent = sender as string;
    }

    private async Task BrowserHasResizedAsync()
    {
        InnerHeight = await BrowserService.GetInnerHeight();
        InnerWidth = await BrowserService.GetInnerWidth();
        svgMapHeight = (int)(InnerHeight * 0.85);
        StateHasChanged();
    }

    private async Task MouseOver(MouseEventArgs e)
    {
        await JsRuntime.InvokeVoidAsync("mouseMoved.showTooltip", e);
    }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            await BrowserHasResizedAsync();
        }

        await JsRuntime.InvokeAsync<object>("browserResize.registerResizeCallback");
        await base.OnAfterRenderAsync(firstRender);
    }

    public void Dispose()
    {
        BrowserService.OnResize -= BrowserHasResizedAsync;
    }
}
```

### functions.js
``` javascript
window.browserResize = {
    getInnerHeight: function () {
        return window.innerHeight;
    },
    getInnerWidth: function () {
        return window.innerWidth;
    },
    registerResizeCallback: function () {
        window.addEventListener("resize", browserResize.resized);
    },
    resized: function () {
        DotNet.invokeMethodAsync('Covid19Tracker', 'OnBrowserResizeEventHandler').then(data => data);
    }
};

window.mouseMoved = {
    showTooltip: function (evt, text) {
        let tooltip = document.getElementById("tooltip");
        tooltip.innerHTML = text;
        tooltip.style.display = "block";
        tooltip.style.left = evt.clientX + 10 + 'px';
        tooltip.style.top = evt.clientY + 10 + 'px';
    },
    hideTooltip: function () {
        var tooltip = document.getElementById("tooltip");
        tooltip.style.display = "none";
    }
};
```

### BrowserService.cs
```C#
using Microsoft.JSInterop;
using System;
using System.Threading.Tasks;

namespace Covid19Tracker.Client.Services
{
    public class BrowserService
    {
        private readonly IJSRuntime js;
        public static event Func<Task> OnResize;
        public static event EventHandler OnMouseOver;

        public BrowserService(IJSRuntime js)
        {
            this.js = js;
        }

        [JSInvokable("Covid19Tracker")]
        public static async Task OnBrowserResizeEventHandler()
        {
            await OnResize?.Invoke();
        }

        [JSInvokable("Covid19Tracker")]
        public static void OnMouseOverEventHandler(string caller)
        {
            OnMouseOver?.Invoke(caller, new EventArgs());
        }

        public async Task<int> GetInnerHeight()
        {
            return await js.InvokeAsync<int>("browserResize.getInnerHeight").ConfigureAwait(false);
        }

        public async Task<int> GetInnerWidth()
        {
            return await js.InvokeAsync<int>("browserResize.getInnerWidth").ConfigureAwait(false);
        }
    }
}
```

### Exception
``` wasm
Uncaught (in promise) Error: System.ArgumentException: There is no loaded assembly with the name 'Covid19Tracker'.
  at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.GetRequiredLoadedAssembly (:5001/Microsoft.JSInterop.Infrastructure.DotNetDispatcher+AssemblyKey assemblyKey) <0x2fce040 + 0x0006e> in <filename unknown>:0 
  at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.ScanAssemblyForCallableMethods (:5001/Microsoft.JSInterop.Infrastructure.DotNetDispatcher+AssemblyKey assemblyKey) <0x2fcd858 + 0x0001c> in <filename unknown>:0 
  at System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].GetOrAdd (:5001/TKey key, System.Func`2[T,TResult] valueFactory) <0x2fca7b0 + 0x00098> in <filename unknown>:0 
  at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.GetCachedMethodInfo (:5001/Microsoft.JSInterop.Infrastructure.DotNetDispatcher+AssemblyKey assemblyKey, System.String methodIdentifier) <0x2fca360 + 0x00066> in <filename unknown>:0 
  at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.InvokeSynchronously (:5001/Microsoft.JSInterop.JSRuntime jsRuntime, Microsoft.JSInterop.Infrastructure.DotNetInvocationInfo& callInfo, Microsoft.JSInterop.Infrastructure.IDotNetObjectReference objectReference, System.String argsJson) <0x2fc9ca0 + 0x0003a> in <filename unknown>:0 
  at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.BeginInvokeDotNet (:5001/Microsoft.JSInterop.JSRuntime jsRuntime, Microsoft.JSInterop.Infrastructure.DotNetInvocationInfo invocationInfo, System.String argsJson) <0x2fc97e8 + 0x00092> in <filename unknown>:0 
    at Object.endInvokeDotNetFromJS (blazor.webassembly.js:1)
    at Object.invokeJSFromDotNet (blazor.webassembly.js:1)
    at _mono_wasm_invoke_js_marshalled (dotnet.3.2.0-preview3.20168.1.js:1)
    at do_icall (:5001/_framework/wasm/dotnet.wasm:wasm-function[6030]:0x10fe48)
    at do_icall_wrapper (:5001/_framework/wasm/dotnet.wasm:wasm-function[1890]:0x52a54)
    at interp_exec_method (:5001/_framework/wasm/dotnet.wasm:wasm-function[1115]:0x2551f)
    at interp_runtime_invoke (:5001/_framework/wasm/dotnet.wasm:wasm-function[5637]:0xf8709)
    at mono_jit_runtime_invoke (:5001/_framework/wasm/dotnet.wasm:wasm-function[5089]:0xdf083)
    at do_runtime_invoke (:5001/_framework/wasm/dotnet.wasm:wasm-function[1405]:0x3da7d)
    at mono_runtime_try_invoke (:5001/_framework/wasm/dotnet.wasm:wasm-function[428]:0xd5ba)
endInvokeDotNetFromJS @ blazor.webassembly.js:1
invokeJSFromDotNet @ blazor.webassembly.js:1
_mono_wasm_invoke_js_marshalled @ dotnet.3.2.0-preview3.20168.1.js:1
do_icall @ dotnet.wasm:1
do_icall_wrapper @ dotnet.wasm:1
interp_exec_method @ dotnet.wasm:1
interp_runtime_invoke @ dotnet.wasm:1
mono_jit_runtime_invoke @ dotnet.wasm:1
do_runtime_invoke @ dotnet.wasm:1
mono_runtime_try_invoke @ dotnet.wasm:1
mono_runtime_invoke @ dotnet.wasm:1
mono_wasm_invoke_method @ dotnet.wasm:1
Module._mono_wasm_invoke_method @ dotnet.3.2.0-preview3.20168.1.js:1
call_method @ dotnet.3.2.0-preview3.20168.1.js:1
(anonymous) @ dotnet.3.2.0-preview3.20168.1.js:1
beginInvokeDotNetFromJS @ blazor.webassembly.js:1
s @ blazor.webassembly.js:1
e.invokeMethodAsync @ blazor.webassembly.js:1
resized @ functions.js:12
Promise.then (async)
resized @ functions.js:12
```

### Further technical details
```
.NET Core SDK (reflecting any global.json):
 Version:   5.0.100-preview.1.20155.7
 Commit:    1c44853056

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19555
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-preview.1.20155.7\

Host (useful for support):
  Version: 5.0.0-preview.1.20120.5
  Commit:  3c523a6a7a

.NET Core SDKs installed:
  3.1.300-preview-015048 [C:\Program Files\dotnet\sdk]
  5.0.100-preview.1.20155.7 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-preview.1.20124.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-preview.1.20120.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-preview.1.20127.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

- Microsoft Visual Studio Community 2019 Preview
   Version 16.6.0 Preview 2.0

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
The one thing which seems wrong to me without much investigation is the usage of the `JSInvokable` attribute in the `BrowserService` class.
The identifier should be unique per class hierarchy, whereas you're passing in same `Covid19Tracker` value.

**Comment by pranavkm** [comment]
```
 resized: function () {
        DotNet.invokeMethodAsync('Covid19Tracker', 'OnBrowserResizeEventHandler').then(data => data);
    }
```

The first argument to `DotNet.invokeMethodAsync` is the assembly name. Have a look at the docs: https://docs.microsoft.com/en-us/aspnet/core/blazor/call-dotnet-from-javascript?view=aspnetcore-3.1#static-net-method-call

**Comment by BrunoBlanes** [comment]
Hi, thanks for the quick reply.
I added `<AssemblyName>Covid19Tracker</AssemblyName>` to my client csproj file and it works fine now. How is it that with Blazor Server it works even without specifying the assembly name on the project file?

About the `JSInvokable` attribute, I had it not set to anything before, but changed to that in a useless try to make it work. I have reverted it now. Shouldn't the compiler let me know if there is more than one attribute of this kind with the same identifier in the same class?
